### PR TITLE
Add support for deferral of player movement, use it to fix two bugs

### LINKED ISSUE #26201: Automine mines wrong tile when crossing between submaps
**Describe the bug**  
Automine mines wrong tile when crossing between submaps

**To Reproduce**  
1. Turn on auto-mine
2. Obtain and wield pickaxe/jackhammer
3. Walk into a mineable tile while crossing between two submaps

**Expected behavior**  
Mine the tile being walked into.

**Actual behavior**  
Some other tile (12 tiles away -- one submap over) gets mined.
automine

If the mined tile is not visible (e.g. you do this at night), sound markers are generated at the mined tile.

**Game Version**: 0.C-33299-gb208fab23e vanilla

### LINKED ISSUE #26239: Auto-move cancelled by door/window opening
**Describe the bug**  
Auto-move is cancelled when encountering a door/window that has to be opened

**To Reproduce**  
Try to auto-move through a closed door or window (includes vehicle doors)

**Expected behavior**  
Auto-moving to selected destination as indicated by planned route.

**Actual behavior**  
Halt after opening door/window

**Screenshots**  
door
window
vehicle

### PR DESCRIPTION
**Pull request by cake-pie**
#### Summary
SUMMARY: Infrastructure "add support for deferral of player movement, use it to fix two bugs"

#### Purpose of change
This adds a bit of machinery on top of existing auto-move code to provide the ability to queue up a player movement for later (e.g. after completing an activity). It is being implemented primarily to facilitate two bugfixes, but it might come in handy for other purposes too.

#### Describe the solution
`player::defer_move()` accepts a tripoint `next` that is adjacent to the player's current position, and sticks it at the front of existing `auto_move_route`, which stores the route for auto-move (left click to go to tile). This leverages existing auto-move functionality, and so the player will automatically move to the `next` location at the next available opportunity.

If the player is not currently in the midst of an auto-move route, `defer_move()` effectively creates a one-tile auto-move route.

If the player is currently in the midst of an auto-move route, then it is required that `next` be adjacent (joins up to) the existing route. The general use case here is basically along these lines: `player::get_next_auto_move_direction()` has been called to [pop the next move](https://github.com/CleverRaven/Cataclysm-DDA/blob/1f2a55c648d49c217c58bc6f1333bbd396b71912/src/player.cpp#L11433-L11434) on the route, but then for some reason we want to defer that move, so we push it back at the front of the route again.

---

### Auto-mine: mine first, then move into mined tile
When auto-mine is on, and player "walks into" a mineable tile while wielding a mining tool, assign the mining activity, defer moving into the target tile, and then return from `game::plmove()` without processing player movement. Auto-move takes care of the movement after mining is completed.

This works for both player-issued tile-by-tile move commands, as well as left-click auto-move routes.

As a bonus, this change makes automining consistent for both passable and impassable tiles -- both now require only one keypress to mine-and-move. Previously, trying to auto-mine into an impassable tile (e.g. wall) would require two keypresses: first to mine, which fails for movement because of the impassable terrain, then again to move into the freshly mined, passable tile.

Fixes #26201 by not moving the player until mining is completed.

#### Describe alternatives you've considered
Make auto-mine not include movement. This would have reduced the utility of auto-mine, forcing mining and moving to be separate keypresses. It would also break the ability to auto-move auto-mine.
Details: https://github.com/CleverRaven/Cataclysm-DDA/issues/26201#issuecomment-429623318

---

### Auto-move through closed doors and windows
(Fixes #26239)

When auto-move encounters a closed door or window, the move command is consumed to perform the open action without actually executing the movement. This causes auto-move to become confused at the next step because the player had failed to advance to the next position that auto-move expected. Auto-move is coded to give up when it encounters such a situation, because it may be caused by the player stumbling due to status effects, etc.

To solve this, we explicitly defer the movement when a door/window is opened while under auto-move, so auto-move doesn't get caught out by the lack of movement.

#### Additional context
Also [remove `on_move_effects();` call](https://github.com/CleverRaven/Cataclysm-DDA/pull/26242/commits/6970603ad64aa565742f7d4d731261db6248b9d9#diff-18513665750ef5adf42b5ec29e14162eL10376) when opening vehicle doors, as it does not involve moving between tiles.

### CODE REVIEW

**Comment by cake-pie** [comment]
While I hammer on the bugfixes, wanted to [CR] and get some opinions on this:

https://github.com/CleverRaven/Cataclysm-DDA/blob/475800bd6349afa051d0da5f48fea4ca0f6e11c1/src/player.cpp#L11460

`auto_move_route` is a vector, and this is a `push_front`. Worth switching to deque?

**Comment by kevingranade** [comment]
> Worth switching to deque?

As we discovered recently in another issue, std::deque is only suitable for large queues (4KB or more!) and should be avoided otherwise.

**Comment by cake-pie** [comment]
All set to go, then.

File: src/player.cpp [hunk 1]
```diff
@@ -11444,6 +11444,24 @@ action_id player::get_next_auto_move_direction()
     return get_movement_direction_from_delta( dp.x, dp.y, dp.z );
 }
 
+bool player::defer_move( tripoint next ) {
+    // next must be adjacent to current pos
+    tripoint dp = next - pos();
+    if( abs( dp.x ) > 1 || abs( dp.y ) > 1 || abs( dp.z ) > 1 ) {
```
Comment (BevapDin) [inline src/player.cpp]: You can use `square_dist` for the adjacent test: `square_dist( ..., pos() ) == 1` means it's adjacent. That's slightly shorter, and you can reduce the second check to one `if` statement.

And your current checks interpret `next == pos()` as adjacent.

Comment (cake-pie) [reply src/player.cpp]: Thanks for the tip, done.

**Comment by kevingranade** [comment]
Very nice solution, I think we might be re-using this a lot as we automate more things.

**Pull request merged**